src/VisualStudio/Core/Def/Implementation/Options/LanguageSettingsPersister.cs

```diff
@@ -40,24 +39,35 @@ internal sealed class LanguageSettingsPersister : ForegroundThreadAffinitizedObj
         /// but the ngen image will exist for the basic map between two reference types, since those are reused.</remarks>
         private readonly IBidirectionalMap<string, Tuple<Guid>> _languageMap;
 
+        private IVsTextManager4 _textManager;
```

Comment:
should I make these volatile? not sure whether await or ConfigureAwait(false) will put memory barrier or not.

Reply:
If this is a reference to an STA object, it shouldn't matter, since you'll only access it from the UI thread anyway. :)

Reply:
I get do GetServiceAsync from BG and then switch to UI thraed to call things that Require UI thread including QI.

in such case, will field require volatile? or one can assume that await or configureAwait make sure memory barrier?

Reply:
As commented below, you'll be on the main thread before assigning this field because even the cast is not free-threaded.
That said, even if you assigned and used this from multiple threads, I've never seen a bug that was due to a lack of a volatile modifier on a field. I'm not an expert on CPU weak memory models, but I suspect any async method's ExecutionContext capture and replay around awaits includes at least one lock and thus a memory barrier.

Reply:
alright, so if one switch to main thread to set field and then switch to BG and use the field. one should be fine. one doesn't need to use LazyInitialize or any of such lib to make sure all threads can observe the changes?

also since it is VS layer where read/write reorder is guaranteed to be preserved, we should be fine right? I was worried that If I keep switch threads, some variable I just set right above, might not observable right below.

Reply:
I don't think the VS layer does anything to make this guarantee, other than the fact that it's an x86 process and thus the CPU is always a strong memory model which probably helps.
I've never seen a bug due to the concern you're raising though (and recognized it as such, anyway).

I think the best place to ask about this is the "Async Discussion" DL internally. I suspect the answer will be that every time an ExecutionContext is applied or captured that a lock has taken place and thus we're free to worry about *other* things.